File: cli/scripts/build-go.mjs
#!/usr/bin/env node
/**
 * Build the Go front controller used as the public loaf command.
 */

import { spawnSync } from "node:child_process";
import { chmodSync, copyFileSync, mkdirSync, readFileSync, writeFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { goBuildArgs } from "./go-build-flags.mjs";

const rootDir = process.cwd();
const launcherSource = join(rootDir, "cli", "runtime", "loaf-launcher.cjs");
const launcherOutput = join(rootDir, "bin", "loaf");

const baseEnv = {
  ...process.env,
  CGO_ENABLED: "0",
  ...pinnedGoToolchainEnv(),
};
const supportedTargets = {
  "darwin-arm64": { goos: "darwin", goarch: "arm64" },
  "darwin-x64": { goos: "darwin", goarch: "amd64" },
  "linux-arm64": { goos: "linux", goarch: "arm64" },
  "linux-x64": { goos: "linux", goarch: "amd64" },
  "win32-arm64": { goos: "windows", goarch: "arm64" },
  "win32-x64": { goos: "windows", goarch: "amd64" },
};
const targets = readBuildTargets(baseEnv);
const dryRun = baseEnv.LOAF_NATIVE_ARTIFACT_DRY_RUN === "1";

mkdirSync(dirname(launcherOutput), { recursive: true });
if (dryRun) {
  console.log(`DRY RUN: would copy Loaf launcher to ${launcherOutput}`);
} else {
  copyFileSync(launcherSource, launcherOutput);
  chmodSync(launcherOutput, 0o755);
  writeFileSync(join(rootDir, "bin", "package.json"), JSON.stringify({ type: "commonjs" }, null, 2) + "\n");
}

for (const target of targets) {
  const nativeName = target.goos === "windows" ? "loaf.exe" : "loaf";
  const nativeOutput = join(rootDir, "bin", "native", target.runtimeID, nativeName);

  if (dryRun) {
    console.log(`DRY RUN: would build ${target.runtimeID} at ${nativeOutput}`);
    continue;
  }

  mkdirSync(dirname(nativeOutput), { recursive: true });
  const result = spawnSync("go", goBuildArgs(nativeOutput, baseEnv), {
    cwd: rootDir,
    env: {
      ...baseEnv,
      GOOS: target.goos,
      GOARCH: target.goarch,
    },
    stdio: "inherit",
  });

  if (result.status !== 0) {
    process.exit(result.status ?? 1);
  }

  chmodSync(nativeOutput, 0o755);
  console.log(`✓ Built Go front controller (${target.runtimeID}): ${nativeOutput}`);
}

if (!dryRun) {
  console.log(`✓ Built Loaf launcher: ${launcherOutput}`);
}

function readBuildTargets(goEnv) {
  const requested = targetListFromEnv(goEnv);
  if (requested.length > 0) {
    return requested.map((runtimeID) => targetFromRuntimeID(runtimeID));
  }
  const current = readCurrentGoTarget(goEnv);
  return [{ ...current, runtimeID: `${nodePlatform(current.goos)}-${nodeArch(current.goarch)}` }];
}

function pinnedGoToolchainEnv() {
  const goMod = readFileSync(join(rootDir, "go.mod"), "utf8");
  const match = goMod.match(/^toolchain\s+(\S+)\s*$/m);
  return match ? { GOTOOLCHAIN: match[1] } : {};
}

function targetListFromEnv(goEnv) {
  return (goEnv.LOAF_BUILD_TARGETS || goEnv.LOAF_NATIVE_TARGETS || "")
    .split(",")
    .map((value) => value.trim())
    .filter(Boolean)
    .filter((value, index, values) => values.indexOf(value) === index);
}

function readCurrentGoTarget(goEnv) {
  const result = spawnSync("go", ["env", "GOOS", "GOARCH"], {
    cwd: rootDir,
    env: goEnv,
    encoding: "utf8",
  });
  if (result.status !== 0) {
    process.stderr.write(result.stderr || "");
    process.exit(result.status ?? 1);
  }
  const [goos, goarch] = result.stdout.trim().split(/\s+/);
  if (!goos || !goarch) {
    console.error("ERROR: could not determine Go target from `go env GOOS GOARCH`");
    process.exit(1);
  }
  return { goos, goarch };
}

function targetFromRuntimeID(runtimeID) {
  const target = supportedTargets[runtimeID];
  if (!target) {
    console.error(`ERROR: unsupported LOAF_BUILD_TARGETS entry ${JSON.stringify(runtimeID)}.`);
    console.error(`Supported targets: ${Object.keys(supportedTargets).join(", ")}`);
    process.exit(1);
  }
  return { runtimeID, ...target };
}

function nodePlatform(goos) {
  if (goos === "windows") return "win32";
  return goos;
}

function nodeArch(goarch) {
  switch (goarch) {
    case "amd64":
      return "x64";
    case "386":
      return "ia32";
    default:
      return goarch;
  }
}

File: internal/devtool/buildgo.go
package devtool

import (
	"fmt"
	"io"
	"os"
	"path/filepath"
	"runtime"
	"strings"
)

// BuildGoOptions drive one native build.
type BuildGoOptions struct {
	RootDir string
	Env     Env
	Runner  Runner
	Stdout  io.Writer
	Stderr  io.Writer
	// Builder compiles one target into dest. Nil means `go build`; tests inject
	// a stand-in that writes fixture bytes.
	Builder func(dest string, target Target, env Env) error
	// HostTarget overrides the platform bin/loaf is copied from. Nil means the
	// running process's GOOS/GOARCH.
	HostTarget *Target
}

// GoBuildArgs is the single source of truth for how the public loaf binary is
// compiled. Every channel (dev, release, verification) must agree with it.
//
// Release metadata is linked via -X main.buildCommit/buildDate only when the
// release workflow sets LOAF_BUILD_COMMIT / LOAF_BUILD_DATE; their absence is
// half of what marks a binary as a dev build (see internal/cli/version.go).
// -buildvcs=true stamps vcs.revision and vcs.modified, which a dev build
// reports as <version>+g<short-sha>[.dirty]. go.mod pins a toolchain that
// stamps linked worktrees; go1.26.6 did not.
func GoBuildArgs(output string, env Env) []string {
	return []string{"build", "-trimpath", "-buildvcs=true", "-ldflags", GoLdflags(env), "-o", output, "./cmd/loaf"}
}

// GoLdflags renders the linker flags for GoBuildArgs.
func GoLdflags(env Env) string {
	parts := []string{"-buildid="}
	if commit := strings.TrimSpace(env["LOAF_BUILD_COMMIT"]); commit != "" {
		parts = append(parts, "-X main.buildCommit="+commit)
	}
	if date := strings.TrimSpace(env["LOAF_BUILD_DATE"]); date != "" {
		parts = append(parts, "-X main.buildDate="+date)
	}
	return strings.Join(parts, " ")
}

// IsReleaseBuild reports whether release metadata is present.
func IsReleaseBuild(env Env) bool {
	return strings.TrimSpace(env["LOAF_BUILD_COMMIT"]) != "" || strings.TrimSpace(env["LOAF_BUILD_DATE"]) != ""
}

// BuildGo compiles the requested targets (LOAF_BUILD_TARGETS, else the host)
// into bin/native/<target>/, staging every binary first so a failing target
// leaves the previously published set intact. bin/loaf becomes a copy of the
// host platform's binary: it is the distribution's entry point, and since the
// Node launcher is gone it is the native binary itself. Builds never retarget
// the development launcher; use Install for explicit activation.
func BuildGo(options BuildGoOptions) error {
	root := options.RootDir
	stdout := options.Stdout
	if stdout == nil {
		stdout = os.Stdout
	}
	env := options.Env.With(map[string]string{"CGO_ENABLED": "0"}).With(PinnedToolchainEnv(root))
	runner := options.Runner
	if runner == nil {
		runner = ExecRunner{}
	}
	builder := options.Builder
	if builder == nil {
		builder = func(dest string, target Target, env Env) error {
			return runner.Run(root, env.With(map[string]string{"GOOS": target.GOOS, "GOARCH": target.GOARCH}), "go", GoBuildArgs(dest, env)...)
		}
	}

	targets, err := requestedTargets(env, root, runner, "LOAF_BUILD_TARGETS", "LOAF_NATIVE_TARGETS")
	if err != nil {
		return err
	}
	host := options.HostTarget
	if host == nil {
		if resolved, err := TargetForGo(runtime.GOOS, runtime.GOARCH); err == nil {
			host = &resolved
		}
	}
	dryRun := env["LOAF_NATIVE_ARTIFACT_DRY_RUN"] == "1"

	nativeRoot := filepath.Join(root, "bin", "native")
	staging := filepath.Join(nativeRoot, ".staging")
	launcher := filepath.Join(root, "bin", "loaf")
	if dryRun {
		fmt.Fprintf(stdout, "DRY RUN: would copy the host binary to %s\n", launcher)
		for _, target := range targets {
			fmt.Fprintf(stdout, "DRY RUN: would build %s at %s\n", target.RuntimeID, filepath.Join(nativeRoot, target.RuntimeID, target.BinaryName()))
		}
		return nil
	}

	if err := os.MkdirAll(filepath.Dir(launcher), 0o755); err != nil {
		return err
	}
	_ = os.RemoveAll(staging)
	defer os.RemoveAll(staging)

	type staged struct{ from, to string }
	var built []staged
	for _, target := range targets {
		dest := filepath.Join(staging, target.RuntimeID, target.BinaryName())
		if err := os.MkdirAll(filepath.Dir(dest), 0o755); err != nil {
			return err
		}
		if err := builder(dest, target, env); err != nil {
			return fmt.Errorf("go build for %s failed: %w", target.RuntimeID, err)
		}
		if err := os.Chmod(dest, 0o755); err != nil {
			return err
		}
		fmt.Fprintf(stdout, "✓ Built Go front controller (%s): %s\n", target.RuntimeID, filepath.Join(nativeRoot, target.RuntimeID, target.BinaryName()))
		built = append(built, staged{from: dest, to: filepath.Join(nativeRoot, target.RuntimeID, target.BinaryName())})
	}
	for _, artifact := range built {
		if err := os.MkdirAll(filepath.Dir(artifact.to), 0o755); err != nil {
			return err
		}
		if err := os.Rename(artifact.from, artifact.to); err != nil {
			_ = os.Remove(artifact.to)
			if err := os.Rename(artifact.from, artifact.to); err != nil {
				return err
			}
		}
	}

	// bin/loaf is the host platform's binary. A build that did not compile the
	// host keeps whatever bin/loaf already exists.
	if host != nil {
		hostBinary := filepath.Join(nativeRoot, host.RuntimeID, host.BinaryName())
		if containsTarget(targets, *host) {
			if err := copyFile(hostBinary, launcher, 0o755); err != nil {
				return fmt.Errorf("publish bin/loaf: %w", err)
			}
			fmt.Fprintf(stdout, "✓ Published Loaf entry point: %s\n", launcher)
		}
	}
	// The plugin shim's `package.json` sibling is gone; keep the old file from
	// lingering in checkouts that predate the change.
	_ = os.Remove(filepath.Join(root, "bin", "package.json"))

	return nil
}

func containsTarget(targets []Target, want Target) bool {
	for _, target := range targets {
		if target.RuntimeID == want.RuntimeID {
			return true
		}
	}
	return false
}

func copyFile(from, to string, mode os.FileMode) error {
	body, err := os.ReadFile(from)
	if err != nil {
		return err
	}
	tmp := to + ".tmp"
	if err := os.WriteFile(tmp, body, mode); err != nil {
		return err
	}
	return os.Rename(tmp, to)
}
